draft-irtf-cfrg-vdaf.md

```diff
@@ -1190,6 +1194,190 @@ how the nonces are chosen, but security requires that the nonces be unique. See
+               host_input_share: bytes):
+    try:
+        host_prep_state = Vdaf.prep_init(
+            vdaf_verify_key,
+            0 if is_leader else 1,
+            agg_param,
+            nonce,
+            public_share,
+            host_input_share,
+        )
+        state CONTINUED(0, host_prep_state)
+    except:
+        state REJECTED
+    return None
+~~~
+
+If the state is `REJECTED`, then processing halts. Otherwise, if the state is
+`CONTINUE`, then processing continues using the following transition rule until
+a terminal state is reached.
+
+## Ping-Pong Topology (Only Two Aggregators)
+
+For VDAFs with precisely two Aggregators (i.e., `Vdaf.SHARES == 2`), the
+following "ping pong" communication pattern can be used. It is compatible with
+any request/response protocol, such as HTTP.
+
+Let us call the initiating party the "Leader" and the responding party the
+"Helper". The high-level idea is that the Leader and Helper will take turns
+running the computation locally until input from their peer is required:
+
+* For a 1-round VDAF (e.g., Prio3 ({{prio3}})), The Leader sends its prep share
+  to the Helper, who computes the prep message locally (the computations are
+  specified in the following sections), computes its output share, then sends
+  the prep message to the Leader. Preparation requires just one request from
+  the Leader to the Helper.
+
+* For a 2-round VDAF (e.g., Poplqr1 ({{poplar1}})), the Leader sends its
+  first-round prep share to the Helper, who replies with the first-round prep
+  message and its second-round prep share. In the next request, the Leader
+  computes its second-round prep share locally, computes its output share, and
+  sends the second-round prep message to the Helper. Finally, the Helper
+  computes its own output share. Preparation requires two requests.
+
+* In general, each request includes the Leader's prep share for the previous
+  round and/or the prep message for the current round; correspondingly, each
+  response consists of the prep message for the current round and the Helper's
+  prep share for the next round.
+
+The Aggregators proceed in this ping-ponging fashion until a step of the
+computation fails, indicating the report is invalid and should be rejected, or
+the last round is reached and each Aggregator computes an output share. All
+told there there are `ceil((Vdaf.ROUNDS+1)/2)` requests sent.
+
+~~~ transition
+State CONTINUED(round_num, host_prep_state)
+
+def ping_pong_prep_next(Vdaf,
+                        is_leader: bool,
+                        agg_param: Vdaf.AggParam,
+                        inbound_type: Optional[str],
+                        inbound: Optional[bytes]):
+    if inbound_type == "initialize":
+        prep_msg, peer_prep_share = None, inbound
+    elif inbound_type == "continue":
+        l = Vdaf.prep_size(agg_param, round_num)
+        prep_msg, prep_share = inbound[:l], inbound[l:]
+        round_num += 1
```

Comment:
@branlwyd to ensure that the round number is computed properly I believe we need increment the counter here. Please double check my math here. (This is a change from https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/pull/461.)

Reply:
Thinking to a large-round VDAF, this round-counting scheme tracks a consistent sequence of rounds between the Leader & Helper; specifically 1→3→5→...

But I think this line should be dropped, so that "continue" messages only increment the round counter once, and the Leader & Helper compute the round sequence 1→2→3→...

The reason I think this is a good change is that the purpose of the round-counter is to track progress/allow for retries at the level of the VDAF-evaluation protocol (e.g. DAP). If the sequence of rounds is 1→3→5→..., then the higher-level protocol has a failure mode if an attempt is made to retry/advance to an even-numbered round; this failure mode does not exist if the round sequence is 1→2→3→...

(I'm not sure we need to include round counting at all in these functions. I'd be happy if these functions defined the sequence of VDAF calls, and left the definition of any retry mechanisms up to the VDAF-evaluation protocol. ISTM that a hypothetical DAP-competitor protocol might define a different retry mechanism, at which point this round-counting is vestigal. Also, these functions are defining a round counter for each report, while even DAP tracks round at the level of an aggregation job, i.e. across many reports.)

Reply:
Right, my goal here was to have `round_num` match the VDAF round, not the ping-pong round. If we go with your suggestion then we'd need to change
```python
if round_num == Vdaf.ROUNDS:
```
to
```python
if round_num = ceil((Vdaf.ROUNDS+1)/2):
```
Does that sound right?

Reply:
Ahhh you're right! What would be useful for DAP would be the "request number" instead of the round number. I started working on this change but found a problem: We now have a callback on `Vdaf` for getting the size of the prep message (`Vdaf.prep_size()`) and unfortunately this needs the round number as input.

Reply:
One way around this would be to length-prefix the prep message in the coalesced `Continue` message type so that we can parse the two. It sort of breaks with the design pattern we've been following for VDAF but I'd be fine with this if you think tracking the request number would be useful.

Reply:
Good call :)

I've removed `Vdaf.prep_size()` and removed the round counter from the state.